lib/truffle/truffle/patching.rb

```diff
@@ -1,34 +1,119 @@
+
+              "rspec/support"                           => :after,
+
+              "unf/normalizer_cruby"                    => :before,
+  }
+
+  PATCHES.each do |file, _|
+    unless File.exist? File.join(DIR, file) + '.rb'
+      raise "file #{file} is missing in #{DIR}"
+    end
+  end
+
+  Dir["#{DIR}/**/*.rb"].each do |file|
+    relative_path = file[(DIR.size + 1)..-4]
+    unless PATCHES.key? relative_path
+      raise "file #{relative_path} is not declared in PATCHES"
+    end
+  end
+
+  def self.find_absolute_path(relative_path)
+    $LOAD_PATH.each do |base|
+      candidate = File.join(base, relative_path) + '.rb'
+      return candidate if File.exist? candidate
+    end
+    raise "absolute path for #{relative_path} was not find in #{$LOAD_PATH}"
+  end
+
+  def self.find_relative_path(absolute_path)
+    absolute_path_with_suffix = if absolute_path.end_with?('.rb')
+                                  absolute_path
+                                else
+                                  absolute_path + '.rb'
+                                end
+
+    raise "file #{absolute_path} does not exist" unless File.exist? absolute_path_with_suffix
+    $LOAD_PATH.each do |base|
+      if absolute_path_with_suffix.start_with?(base)
+        return File.basename absolute_path_with_suffix[(base.size)..-4]
+      end
+    end
+
+    return nil
+  end
+
+  def self.required?(relative_path)
+    $LOADED_FEATURES.include? find_absolute_path relative_path
+  end
 end
 
 module Kernel
 
   private
 
-  alias_method :require_without_truffle_patching, :require
+  alias_method :require_without_truffle_patching, :gem_original_require
 
-  def require(path)
-    required = begin
-      require_without_truffle_patching path
-    rescue LoadError
-      if Truffle::Patching::TRUFFLE_PATCHES[path]
-        true # Pretend the file required so we attempt patching with something that should avoid the LoadError (e.g., replacing code using C extensions).
-      else
-        raise
-      end
-    end
+  def gem_original_require(path)
+    extname        = File.extname path
+    path_no_suffix = path[0..(-1-extname.size)]
+    relative_path  = if path_no_suffix.start_with?('/')
+                       Truffle::Patching.find_relative_path path_no_suffix
+                     else
+                       path_no_suffix
+                     end
 
-    if required && Truffle::Patching::TRUFFLE_PATCHES[path]
-      Truffle::System.log :PATCH, "applying #{path}"
-      require_without_truffle_patching "#{Truffle::Patching::TRUFFLE_PATCHES_DIRECTORY}/#{path}.rb"
-    end
+    operation = Truffle::Patching::PATCHES[relative_path]
+
+    # not patched requiring normally
+    return require_without_truffle_patching path if operation.nil?
 
-    required
+    # apply patch based on operation
+    log = -> { Truffle::System.log :PATCH, "#{operation} original requiring '#{path}'" }
+    return case operation
```

Comment:
It's intentional, to draw readers attention to that the last value in a branch is intended return value. The returns could be in each branch but that is quite repetitive. There is no wrapped line so the indentation is not an issue.